### PR TITLE
Draft: Add type annotations to mobject.py

### PR DESCRIPTION
**Pull request by Timmmm**
**This is a draft - it's not quite finished yet; I just put it up for discussion (and in case I never finish it and somebody wants to take over). Or in case you just want to land it unfinished.**

This adds quite a few type annotations to mobject.py (and a few to other files to help). Type checked with Pyright - MyPy basically can't handle this at all. There are quite a few type errors left but way fewer than when I started (Pyright gave up about half way through the file!)

Some misc notes:

* The code is not clear where `Sequence[float]` is allowed and where `np.ndarray` is (or either). I've mostly just put `np.ndarray`.
* Similarly for `Color` vs `str`. Not clear where either is allowed and some of the existing type annotations seem to be wrong.
* Changed `np.float64` to `float` for simplicity. They're officialy not compatible but they actually are the same.
* Removed unused `axes` parameter from `rotate_about_origin`.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -1,3 +1,7 @@
+from __future__ import annotations
+
+from _typeshed import SupportsLessThan
```
Comment (Timmmm) [inline manim/mobject/mobject.py:3]: Probably should remove this actually. Rather unfortunate that Python doesn't care enough about type correctness to put type annotations on the standard library.

**Comment by Darylgolden** [comment]
Just a comment about writing unions as `X | Y` - since this is only going to be introduced with Python 3.10, and we're officially supporting Python 3.7+, you should use the old syntax.

**Comment by GameDungeon** [comment]
This is similar to a pull I already made (but you did better), where I made a dedicated file for this. I think the changes should be combined: https://github.com/ManimCommunity/manim/pull/2052/

**Comment by GameDungeon** [comment]
It seems to really not like your _typeshed

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -14,21 +18,11 @@
...
-    Tuple,
-    Type,
-    TypeVar,
-    Union,
-)
+from typing import TYPE_CHECKING, Any, Callable, Mapping, Sequence, TypeVar, Union, cast
 
 import numpy as np
+
+# TODO(types): colour has no type annotations.
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:25]: Would be fixed by either #1972 or #2050

File: manim/mobject/mobject.py [hunk 3]
```diff
@@ -78,20 +93,44 @@ class Mobject:
 
     """
 
+    color: Color | ColorStr | None
+    name: str
+    dim: int
+    target: Mobject | None
+    z_index: float
+    point_hash: int | None
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:101]: Like already mentioned, those are not supported yet. Use Union.

File: manim/mobject/mobject.py [hunk 4]
```diff
@@ -464,15 +506,15 @@ def add_to_back(self, *mobjects: "Mobject") -> "Mobject":
             raise ValueError("A mobject shouldn't contain itself")
 
         for mobject in mobjects:
-            if not isinstance(mobject, Mobject):
+            if not isinstance(mobject, Mobject):  # type: ignore
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:509]: Why is this here?

File: manim/mobject/mobject.py [hunk 5]
```diff
@@ -499,13 +541,13 @@ def remove(self, *mobjects: "Mobject") -> "Mobject":
                 self.submobjects.remove(mobject)
         return self
 
-    def __sub__(self, other):
+    def __sub__(self, other: Mobject) -> Mobject:
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:544]: Lying about the return value

File: manim/mobject/mobject.py [hunk 6]
```diff
@@ -499,13 +541,13 @@ def remove(self, *mobjects: "Mobject") -> "Mobject":
...
         raise NotImplementedError
 
-    def __isub__(self, other):
+    def __isub__(self, other: Mobject) -> Mobject:
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:547]: Lying about the return value

File: manim/mobject/mobject.py [hunk 7]
```diff
@@ -1308,44 +1373,35 @@ def apply_points_function_about_point(
...
         # Redundant with default behavior of scale now.
         return self.scale(scale_factor, **kwargs)
 
-    @deprecated(
-        since="v0.11.0",
-        until="v0.12.0",
-        replacement="scale",
-    )
-    def scale_about_point(self, scale_factor, point):
+    def scale_about_point(self: MOS, scale_factor: float, point: np.ndarray) -> MOS:
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:1386]: Why did you remove the depreciation?

File: manim/mobject/mobject.py [hunk 8]
```diff
@@ -1437,19 +1505,25 @@ def is_off_screen(self):
...
+    ) -> MOS:
         return self.stretch(factor, dim, about_point=point)
 
-    @deprecated(
-        since="v0.11.0",
-        until="v0.12.0",
-        replacement="stretch",
-    )
-    def stretch_in_place(self, factor, dim):
+    def stretch_in_place(self: MOS, factor: float, dim: int) -> MOS:
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:1516]: Same here

File: manim/mobject/mobject.py [hunk 9]
```diff
@@ -51,11 +44,33 @@
...
+    from ..camera.camera import Camera
+    from ..utils.bezier import Interpolable
+
+# Generic type
+T = TypeVar("T")
+
+Updater = Union[Callable[["Mobject"], None], Callable[["Mobject", float], None]]
+# Any subclass of Mobject. Most Mobject methods `return self` so this is used
+# so that e.g. Circle.scale(2) returns a Circle, not an Mobject.
+MOS = TypeVar("MOS", bound="Mobject")
```
Comment (GameDungeon) [inline manim/mobject/mobject.py:60]: Maybe a better name for this? I mean it's not needed. Maybe MOB instead of MOS

**Review by GameDungeon — CHANGES_REQUESTED** [review]

File: manim/mobject/mobject.py (hunk 2 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:25]: I don't think so. This is about [this colour](https://pypi.org/project/colour/) not the Manim `Color` type.

File: manim/mobject/mobject.py (hunk 4 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:509]: The type annotations already say that `mobject` should be an `Mobject` so this is an unnecessary `isinstance()` and Pyright gives an error about it.

However I left the check in because not everyone uses type annotations and presumably this runtime check was helpful in the past.

File: manim/mobject/mobject.py (hunk 5 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:544]: Do you mean the return type is wrong? If so what should it be? It's pretty hard to tell in some cases!

I don't think you mean *lying* - "lying" means being deliberately deceptive and I hope you aren't accusing me of that!

File: manim/mobject/mobject.py (hunk 7 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:1386]: Weird, I definitely didn't do that! Maybe an autoformatter gone wrong? Good spot though, I'll add it back.

File: manim/mobject/mobject.py (hunk 9 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:60]: "MOS" was meant to be "MObject Subclass"...  Maybe that was not obvious - happy to change it! Why MOB? Just MOBject?

**Comment by Timmmm** [comment]
> Just a comment about writing unions as `X | Y` - since this is only going to be introduced with Python 3.10, and we're officially supporting Python 3.7+, you should use the old syntax.

Actually I'm pretty sure `from __future__ import annotations` takes care of that. I don't have Python 3.7 to hand but it works fine in Python 3.8:

```
~ python3
Python 3.8.7 (default, Jan  9 2021, 01:55:12) 
[Clang 9.0.0 (tags/RELEASE_900/final)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x: int | None = 5
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
>>> from __future__ import annotations
>>> x: int | None = 5
>>> x
5
```

I think it doesn't work in some contexts, but Pyright seems to know the rules and give an error when you get it wrong. Do you have tests that run it with Python 3.7 just to be sure?

File: manim/mobject/mobject.py (hunk 6 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:547]: > This is similar to a pull I already made (but you did better), where I made a dedicated file for this. I think the changes should be combined: #2052

Nice, I had a look and I think our changes don't conflict much (if at all) so hopefully we can land them separately.

File: manim/mobject/mobject.py (hunk 2 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:25]: Though you have just made me realise that I put `Color` in loads of places where I think it should have been `Colors`. I'll fix it.

**Comment by Timmmm** [comment]
> I think it doesn't work in some contexts

Yeah so the thing you *can't* do until Python 3.10 (even with `from __future__ import annotations`) is use `|` in a non type-hint context. In other words this is fine:

    x: int | str = 5

But this isn't

    IntOrStr = int | str

Pyright warns you an error for the second case.

File: manim/mobject/mobject.py (hunk 4 above)
Comment (Timmmm) [reply manim/mobject/mobject.py:528]: I added a comment to explain it.

**Comment by GameDungeon** [comment]
> I think it doesn't work in some contexts, but Pyright seems to know the rules and give an error when you get it wrong. Do you have tests that run it with Python 3.7 just to be sure?

I mean look at our CI, we do have tests

**Comment by Timmmm** [comment]
> I mean look at our CI, we do have tests

Alright, I was just asking. No need to get defensive!

**Comment by Darylgolden** [comment]
Hi! Are you still interested in working on this?

**Comment by Darylgolden** [comment]
I'll close this for now, please feel free to reopen if you intend to continue working on this!
